packages/hdwallet-vultisig/src/bitcoin.ts

```diff
@@ -0,0 +1,172 @@
+  if (!bytes) throw new Error("Invalid hex string");
+
+  return Uint8Array.from(bytes.map((byte) => parseInt(byte, 16)));
+};
+
+async function addInput(psbt: bitcoin.Psbt, input: core.BTCSignTxInput): Promise<void> {
+  switch (input.scriptType) {
+    // Phantom supposedly supports more scriptTypes but in effect, doesn't (currently)
+    // https://github.com/orgs/phantom/discussions/173
+    case BTCInputScriptType.SpendWitness: {
+      psbt.addInput({
+        hash: input.txid,
+        index: input.vout,
+        nonWitnessUtxo: Buffer.from(input.hex, "hex"),
+      });
+
+      break;
+    }
+    default:
+      throw new Error(`Unsupported script type: ${input.scriptType}`);
+  }
+}
+
+async function addOutput(
+  wallet: core.BTCWallet,
+  psbt: bitcoin.Psbt,
+  output: core.BTCSignTxOutput,
+  coin: string
+): Promise<void> {
+  if (!output.amount) throw new Error("Invalid output - missing amount.");
+
+  const address = await (async () => {
+    if (output.address) return output.address;
+
+    if (output.addressNList) {
+      const outputAddress = await wallet.btcGetAddress({ addressNList: output.addressNList, coin, showDisplay: false });
+      if (!outputAddress) throw new Error("Could not get address from wallet");
+      return outputAddress;
+    }
+  })();
+
+  if (!address) throw new Error("Invalid output - no address");
+
+  psbt.addOutput({ address, value: BigInt(output.amount) });
+}
+
+export async function bitcoinSignTx(
+  wallet: core.BTCWallet,
+  msg: core.BTCSignTx,
+  provider: VultisigUtxoProvider
+): Promise<core.BTCSignedTx | null> {
+  try {
+    const network = getNetwork(msg.coin);
+    const psbt = new bitcoin.Psbt({ network });
+
+    psbt.setVersion(msg.version ?? 2);
+    if (msg.locktime) {
+      psbt.setLocktime(msg.locktime);
+    }
+
+    for (const input of msg.inputs) {
+      await addInput(psbt, input);
+    }
+
+    for (const output of msg.outputs) {
+      await addOutput(wallet, psbt, output, msg.coin);
+    }
+
+    if (msg.opReturnData) {
+      const data = Buffer.from(msg.opReturnData, "utf-8");
+      const embed = bitcoin.payments.embed({ data: [data] });
+      const script = embed.output;
+      if (!script) throw new Error("unable to build OP_RETURN script");
+      // OP_RETURN_DATA output is always 0 value
+      psbt.addOutput({ script, value: BigInt(0) });
+    }
+
+    const inputsToSign = await Promise.all(
+      msg.inputs.map(async (input, index) => {
+        const address = await wallet.btcGetAddress({
+          addressNList: input.addressNList,
+          coin: msg.coin,
+          showDisplay: false,
+        });
+
+        if (!address) throw new Error("Could not get address from wallet");
+
+        return {
+          address,
+          signingIndexes: [index],
+          sigHash: bitcoin.Transaction.SIGHASH_ALL,
+        };
+      })
+    );
+
+    const signedPsbtHex = await provider.signPSBT(fromHexString(psbt.toHex()), { inputsToSign }, false);
```

Comment:
also wondering if we can just get the buffer value and get rid of the `fromHexString` alltogether.
```suggestion
    const signedPsbtHex = await provider.signPSBT(psbt.toBuffer()), { inputsToSign }, false);
```